mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
   extends RegressionModel[Vector, LinearRegressionModel]
   with LinearRegressionParams {
 
+  @transient private var trainingResults: Option[LinearRegressionTrainingResults] = None
+
+  /**
+   * Gets results (e.g. residuals, mse, r^2) of model on training set. This method should only
+   * be called on the driver (it is not available on workers).
+   */
+  def getTrainingResults: Option[LinearRegressionTrainingResults] = trainingResults
```

Comment:
Let's stick with the design doc naming, or modify the design doc if you think other naming is better.  I like the more succinct usage in the doc:
```
lrModel.summary.objectiveTrace
```
That will require summary to throw an exception if the results are not available, but that seems OK to me as long as we make it a meaningful exception.  We could also provide:
```
def hasSummary: Boolean
def setSummary(summary: LinearRegressionSummary): this.type
def getSummary // for convenience, if you want
```

I also like calling it Summary instead of TrainingResults since it's shorter and may be more recognizable to R users.